Bugfix/add relay indicator to capi commads

morantr opened the pull request:
Add relay indicator to CAPI commands with message types that require it according to Table 6-4 on IEEE 1905.1-2013, to solve #1011.

As part of this PR, changed also other messages that should have been sent as relayed multicast, to relayed multicast.  

**Help Needed!**
For some reason sending messages with relay indicator casing to some other message to not be received. For example, when the repeater sends an auto-config search message (relay=1), the controller response to it (relay=0) but the response is not being received on the repeater side.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3748,8 +3748,10 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
-    return message_com::send_cmdu(master_socket, cmdu_tx, backhaul_params.controller_bridge_mac,
-                                  backhaul_params.bridge_mac);
+
+    static const std::string MULTICAST_MAC_ADDR = "01:80:c2:00:00:13";
```
Can't we use transport_socket_thread::MULTICAST_MAC_ADDR here?

tomereli replied:
I guess not since it is defined in the BTL, and the son slave does not link with it.
@morantr - please add this to `network_utils` instead - you should name the variable `MULTICAST_1905_MAC_ADDR` since its the 1905 multicast mac according to the 1905.1 specification.

morantr replied:
done, although it wouldn't be necessary if we had a unified Agent.
I'm saying it a lot, do I?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3748,8 +3748,10 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
-                                  backhaul_params.bridge_mac);
+
+    static const std::string MULTICAST_MAC_ADDR = "01:80:c2:00:00:13";
+    auto dst_addr = multicast ? MULTICAST_MAC_ADDR : backhaul_params.controller_bridge_mac;
```
Since multicast is never used for VS messages, there is nothing in common between the multicast and non-multicast case. So I think it's better to create a separate function send_cmdu_multicast. (nitpick)

morantr replied:
not critical IMO if it is a default argument or a different function.
Anyway, when we will move to a unified Agent, we send it directly to the bus. So this function will look differently.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1926,10 +1926,13 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
             return false;
         }
+        cmdu_header->flags().relay_indicator = true;
 
```
(Not for this PR) Maybe it would be better to move this code entirely to backhaul_manager. Indeed, if we just forward the notification as-is to backhaul manager, we can handle everything there. backhaul_manager anyway needs to handle the notification since it maintains a cache of the state of all radios together.

morantr replied:
Or... move the code from the backhaul manager to the unified Agent...

arnout reviewed (approved):
Looks very good - except that CI fails. From the log files, I don't immediately see what the issue is. Is this the problem we talked about on zoom a couple of days ago? If you want, I'm willing to debug it.

tomereli commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -350,6 +350,23 @@ bool beerocks_ucc_listener::get_send_1905_1_tlv_hex_list(
     return false;
 }
 
+static void set_relay_indicator(std::shared_ptr<ieee1905_1::cCmduHeader> &cmdu_header)
```
This should be done directly in `CmduMessageTx::create`, not here in the listener, so we are sure we don't ever forget to set the relay indicator correctly, not only in the UCC listener. This way you will also take care of the destination (multicast) mac address transparently instead of having to manually set it whenever you create such a CMDU.

morantr replied:
Good idea.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3728,7 +3728,7 @@ bool slave_thread::ap_manager_heartbeat_check()
 }
 
-bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
+bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx, bool multicast)
```
This should be done automatically - why do you need a parameter? You can check the message type and use multicast if the message type is one of the messages defined in Table 6-4.

morantr replied:
Good idea.

arnout commented:
I couldn't reproduce the CI failure locally so I rebased on master and pushed again...

morantr committed:
framework: Set relay indicator automatically

Add relay indicator automatically to messages on CMDU creation for
message types that require it according to Table 6-4 on IEEE 1905.1-2013.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Move multicast mac address to network_utils and rename it

Preparative commit.
Move MULTICAST_MAC_ADDR from BTL to network_utils and rename it to "MULTICAST_1905_MAC_ADDR".
Update all places which use it to the new namespace and name.

This is needed since that on the follow-up commit it is required to use this address from son_slave,
which doesn't have access to BTL.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Allow sending a multicast message from son_slave

Add a possibility to send multicast messages from son_slave.
Specifically, the son_slave sends Topology Notification that needs to be multicast
according to IEEE 1905.1 standard, Table 6-4.
On the on of the follow-up commits this message will be sent on multicast.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backaul_manager: Allow sending messages from son_slave to multicast addr

Currently, if a message coming from son_slave has a multicast address, the backhaul manager
ignores it and sends the message to the controller mac.
Replace the fixed controller destination where we send the CMDU coming from son_slave, to
the address which the son_slave has requested to send the message.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -687,7 +687,8 @@ void beerocks_ucc_listener::handle_wfa_ca_command(const std::string &command)
             // use TLV data from the command (if exists)
             m_cmdu_tx.reset();
-            if (!cmdu_tx.create(g_mid, ieee1905_1::eMessageType(message_type))) {
+            auto cmdu_header = cmdu_tx.create(g_mid, ieee1905_1::eMessageType(message_type));
```
This change is no longer needed.

morantr replied:
removed

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3748,8 +3748,12 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
-                                  backhaul_params.bridge_mac);
+
+    auto dst_addr =
+        cmdu_tx.getMessageType() == ieee1905_1::eMessageType::TOPOLOGY_NOTIFICATION_MESSAGE
```
Shouldn't we send with multicast to more message types?

morantr replied:
not from son_slave.

tomereli reviewed (approved):
some small comments but looks great!

arnout commented:
I suspect we are seeing the combination of a number of issues in the failing test...

- Since we are using pcap to capture packets (#30, #1090), we receive outgoing packets as well as incoming ones. For normal packets this doesn't happen because they are sent on the bridge and we don't sniff the bridge. For relayed multicast packets, however, we send it to each individual interface. Thus, whenever a relayed multicast packet is sent, it is also received 3 times (wlan0, wlan2, eth0).
- Since we don't have an ebtable rule to disable bridge handling of ieee1905.1 packets, the br-lan bridge will relay the packets even if the relayed multicast bit is not set. Thus, every relayed multicast packet will be received twice on each interface.
- Because of the combination of the above two, it is possible that the first time the transport receives a packet, it came from one of the dummy interfaces (wlan0, wlan2) instead of the interface it really came from (eth0). Therefore, it will be relayed back again to eth0 and return to the sender.
- We send to autoconfig search messages back-to-back, with a different MID.
- The transport de-duplication only seems to keep track of the last received MID. Thus, when the second autoconfig search has been received, and then a duplicate of the first one is received, it will be considered as a new one so it will be relayed again.

The combination of all of the above leads to a rush of duplicate packets coming in into the transport, which overflows the pcap buffer and thus the autoconfig response message is lost. The autoconfig response is unicast and is not retransmitted, so it really only arrives once in the repeater, so it is easy to loose it between all the rest.

I think fixing any one of the above issues will fix the problem. E.g. not setting the relay bit will fix it because only the bridge will do relaying, not the transport.

I think the easiest fix will be to set the ebtables rule in the containers so the bridge itself doesn't do any relaying: `ebtables -A FORWARD  -d 01:80:c2:00:00:13 -j DROP`.

morantr committed:
scripts: Add ebtables rule

When setting the relay indicator, it is observed that some message does not get to their
destination.
As it is mentioned on https://github.com/prplfoundation/prplMesh/pull/1029#issuecomment-608353274
The reason for that is a result of a combination of several issues:
1. Since we are using pcap to capture packets (#30, #1090), we receive outgoing packets as well
as incoming ones. For normal packets, this doesn't happen because they are sent on the bridge
and we don't sniff the bridge. For relayed multicast packets, however, we send it to each individual
interface. Thus, whenever a relayed multicast packet is sent, it is also received 3 times
(wlan0, wlan2, eth0).
2. Since we don't have an ebtable rule to disable bridge handling of ieee1905.1 packets, the br-lan
bridge will relay the packets even if the relayed multicast bit is not set. Thus, every relayed
multicast packet will be received twice on each interface.
3. Because of the combination of the above two, it is possible that the first time the transport
receives a packet, it came from one of the dummy interfaces (wlan0, wlan2) instead of the
interface it really came from (eth0). Therefore, it will be relayed back again to eth0 and
return to the sender.
3. We send to autoconfig search messages back-to-back, with a different MID.
4. The transport de-duplication only seems to keep track of the last received MID.
Thus, when the second autoconfig search has been received, and then a duplicate of the
first one is received, it will be considered as a new one so it will be relayed again.

The combination of all of the above leads to a rush of duplicate packets coming into
the transport, which overflows the pcap buffer and thus the autoconfig response message
is lost. The autoconfig response is unicast and is not retransmitted, so it really only arrives
once in the repeater, so it is easy to lose it between all the rest.

This commit adds ebtable rule that so the bridge itself doesn't do any relaying.
In order to use ebtable on docker, add it to the docker file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr commented:
Thanks @arnout I did it on our server and it worked. Added one more commit that fixes it.
I had to update the docker image on our machine to make work. So I guess it still fail on gitlab is since we need to update on the gitlab server too. @rmelotte can you help with that?
